subsys/zigbee/cli/zigbee_cli_cmd_zcl_groups.c

```diff
@@ -0,0 +1,996 @@
+				packet_info->prof_id,
+				ZB_ZCL_DISABLE_DEFAULT_RESPONSE,
+				packet_info->cb);
+}
+
+/**@brief Handles timeout error and invalidates group command entry in context manager.
+ *
+ * @param[in] index   Index of context entry with group command data to invalidate.
+ */
+static void zcl_groups_cmd_timeout(zb_uint8_t index)
+{
+	struct ctx_entry *entry = ctx_mgr_get_entry_by_index(index);
+
+	if (entry == NULL) {
+		LOG_ERR("Couldn't get entry %d - entry not found", index);
+		return;
+	}
+
+	zb_cli_print_error(entry->shell, "Request timed out", ZB_FALSE);
+
+	ctx_mgr_delete_entry(entry);
+}
+
+/**@brief Function to send the ZCL groups cmds.
+ *
+ * @param entry   Pointer to the entry with frame to send in the context manager.
+ *
+ * @return Error code with function execution status.
+ */
+static int zcl_groups_cmd_send(struct ctx_entry *entry)
+{
+	zb_ret_t zb_err_code;
+	uint8_t entry_index = ctx_mgr_get_index_by_entry(entry);
+
+	zb_err_code = ZB_SCHEDULE_APP_ALARM(zcl_groups_cmd_timeout, entry_index,
+					    (CMD_ENTRY_TIMEOUT_S * ZB_TIME_ONE_SECOND));
+	if (zb_err_code != RET_OK) {
+		zb_cli_print_error(entry->shell, "Couldn't schedule timeout cb.", ZB_FALSE);
+		goto cmd_send_error;
+	}
+
+	zb_err_code = ZB_SCHEDULE_APP_CALLBACK(entry->groups_cmd_data.send_fn, entry_index);
+
+	if (zb_err_code != RET_OK) {
+		zb_cli_print_error(entry->shell, "Can not schedule zcl frame.", ZB_FALSE);
+
+		zb_err_code = ZB_SCHEDULE_APP_ALARM_CANCEL(zcl_groups_cmd_timeout, entry_index);
+		if (zb_err_code != RET_OK) {
+			zb_cli_print_error(entry->shell, "Unable to cancel timeout timer", ZB_TRUE);
+		}
+		goto cmd_send_error;
+	}
+
+	return 0;
+
+cmd_send_error:
+	/* Make sure ZBOSS buffer API is called safely. */
+	zb_osif_disable_all_inter();
+	zb_buf_free(entry->generic_cmd_data.packet_info.buffer);
+	zb_osif_enable_all_inter();
+
+	/* Invalidate an entry with frame data in the context manager. */
+	ctx_mgr_delete_entry(entry);
+
+	return -ENOEXEC;
+}
+
+/**@brief Function called when command is APS ACKed - prints done and invalidates groups command
+ *        entry in context manager.
+ *
+ * @param[in] bufid   Reference to a ZBOSS buffer containing zb_zcl_command_send_status_t data.
+ */
+static void zb_zcl_cmd_acked(zb_uint8_t bufid)
+{
+	uint8_t index = 0;
+	zb_ret_t zb_err_code = RET_OK;
+	zb_zcl_command_send_status_t *cmd_status = NULL;
+	struct ctx_entry *entry = ctx_mgr_get_entry_by_index(index);
+
+	if (bufid == ZB_BUF_INVALID) {
+		return;
+	}
+
+	cmd_status = ZB_BUF_GET_PARAM(bufid, zb_zcl_command_send_status_t);
+
+	/* Handle only success status. */
+	if (cmd_status->status != RET_OK) {
+		goto exit;
+	}
+
+	/* Find entry of CTX_MGR_GROUPS_CMD_ENTRY_TYPE type with matching buffer id.
+	 * When ZCL command is APS ACKed, callback is called with the same buffer id.
+	 * This is used to find matching ctx entry.
+	 */
+	while (entry != NULL) {
+		if ((entry->type != CTX_MGR_GROUPS_CMD_ENTRY_TYPE) || (entry->taken != true)) {
+			continue;
```

Comment:
Is not it going to be an infinite loop?

Reply:
It won't because  we `index++` and `entry = ctx_mgr_get_entry_by_index(index)` and NULL is returned after an attempt to get entry with index outside the table.

This takes advantage of how ZBOSS works with finding an entry that corresponds to the buffer the request was sent with.
Also, during ZIgbee shell refactoring it would be nice to make a function to do this `find entry by buffer` and should work fine once the zcl_data structure is present in ctx_mgr, similarly to zdo_data.

Reply:
What I meant - if you `continue` here (break the execution of the loop body), neither the index nor the entry will be updated, so you will be checking the same entry infinitely. Am I missing something?

Reply:
You're right, this will end up in infinite loop. Will fix that